spring-cloud-gcp-starters/spring-cloud-gcp-starter-sql/src/main/java/org/springframework/cloud/gcp/sql/autoconfig/GcpCloudSqlAutoConfiguration.java

```diff
@@ -113,34 +111,44 @@ public CloudSqlJdbcInfoProvider defaultJdbcInfoProvider() {
 	}
 
 	@Bean
-	@ConditionalOnMissingBean(DataSource.class)
-	public DataSource cloudSqlDataSource(CloudSqlJdbcInfoProvider cloudSqlJdbcInfoProvider)
-			throws GeneralSecurityException, IOException, ClassNotFoundException {
-		HikariConfig hikariConfig = new HikariConfig();
-		hikariConfig.setDriverClassName(cloudSqlJdbcInfoProvider.getJdbcDriverClass());
-		hikariConfig.setJdbcUrl(cloudSqlJdbcInfoProvider.getJdbcUrl());
-		hikariConfig.setUsername(this.gcpCloudSqlProperties.getUserName());
-		hikariConfig.setPassword(this.gcpCloudSqlProperties.getPassword());
-		// Setting this is useful to disable connection initialization. Especially in unit tests.
-		hikariConfig.setInitializationFailFast(this.gcpCloudSqlProperties.isInitFailFast());
-
-		return new HikariDataSource(hikariConfig);
+	@Primary
+	public DataSourceProperties cloudSqlDataSourceProperties(DataSourceProperties properties,
```

Comment:
Why can't we just *mutate* injected `DataSourceProperties` instead of creating a new bean?
I understand that you are going to override that one for the `DataSourceAutoConfiguration` via `@Primary`, but this `@Bean` will make two of `DataSourceProperties` beans in the application meanwhile one of them becomes obsolete anyway.

Reply:
good point! is there a way to let spring call my method that has no return value, but still inject DataSourceProperties into the param?  E.g.,

@ Some annotation?
public void overrideCloudSqlDataSourceProperties(DataSourceProperties properties) {
 ...
}

Reply:
Well, I think need a separate inner `@Configuration` class with the `DataSourceProperties` and `CloudSqlJdbcInfoProvider` ctor injections and make an appropriate modification there.

Reply:
i've tried a couple of inner autoconfigs - but unless DataSourceProperties is returned as a bean, the values are not kept.  Returning a bean seams to be the only way. WDYT if we clone the properties and return a new instance?